### DISCUSSION TITLE
rspamadm template --lua-env

### DISCUSSION THREAD

**Discussion by TonyGravagno** [rspamd/rspamd#3771] [Q&A]
I've been studying the doc to understand some details about [Jinja templating in config files](https://rspamd.com/doc/configuration/index.html#jinja-templating). I'm going to fix the text to be more explicit:

From
> by specifying additional environment files with the --lua-env command line argument.

To
> by specifying additional environment files with the rspamadm utility,
> using the 'template' command with the '--lua-env' command line option.

This will also link to the [Rspamadm command](https://rspamd.com/doc/rspamadm.html) page, which shows 'template' as an available command. But there is no other information on that page about 'template'. The doc should include info about options like '--lua-env' and '--suffix', as seen in the [template.lua code](https://github.com/rspamd/rspamd/blob/master/lualib/rspamadm/template.lua#L20). On confirmation I'll add the info from the original [v1.9.1 release notes](https://rspamd.com/announce/2019/04/05/rspamd-1.9.1.html).

Questions:

1) Can someone post an example of an rspamadm command along with a related template? For example:  
`rspamadm template --lua-env /x1/x2/configfile.lua` << is that correct syntax?

2) Can someone post a real-life example of such a configfile.lua? From the doc, I'm understanding that once we create a .lua file, like the example below, and execute the above rspamadm CLI, that the environment will contain var1, subvar1, etc.: 
 
```lua
return {
  var1 = "value",
  var2 = {
    subvar1 = "foorath
    subvar2 = true,
  }
}
```

3) The doc continues, saying with those environment vars in place, we can then use them in .conf files with embedded Jinja templating:

```jinja
{% if env.var2.subvar2 %}
foo = {= env.var1 =};
baz = {= env.var2.subvar2 =};
{% endif %}
```

Can someone post a real-life example of such a config file with Jinja template syntax?

4) I'm confused about timing. We have a running rspamd installation. We create the lua code to set the environment. We run rspamadm once from shell. Then we restart rspamd. Is that the workflow?
So to change those environment vars on a schedule, use cron to execute rspamadm, then reload ( versus restart ) the rspamd service? (See below before answering)

5) The doc says : 
> The specified Lua program file will be read by the root user if the Rspamd main process starts as root and then drops privileges.

To be clear, is that saying we need to sudo rspamadm, or run from root crontab? Why does that say "and then drops privileges"? I'm thinking it's not relevant that privileges are dropped after startup, we assume that this happens. I think this is just telling us that the environment is preserved even when the privileges change?

Coming back to question 4, I think this is saying the rspamd service is started as root, so the environment must be established in root, and that means a simple 'reload' may not work, we may actually need to restart the service if there is an environment change.

6) How is the environment data persisted? That is, if we run the above rspamadm to set the environment with Lua, restart the rspamd service, then restart the system, is the environment lost? Where would it be persisted? Or do we need to ensure that the environment is re-established with a call to rspamadm before every rspamd restart?

7) Ultimately, I'm not understanding why we would run Lua code to set the environment before processing config files. Why not invoke the Lua code from the config files? Examples would help to explain the use case that resulted in this functionality being added in v1.9

Thanks for your time and patience!

**Comment by fatalbanana** [comment]
0. To use a Lua environment file rspamd itself should be started with the `--lua-env /path/to/environment` argument. To correctly render config files that use these variables `rspamadm template`  should have this flag too. Use of environment variables rather than Lua is by my estimation probably most convenient for most users (but Lua is more flexible).

1. This command wants a file to apply templating to so something like `rspamadm template --lua-env /path/to/environment /etc/rspamd/local.d/worker-controller.inc` for e.g.

2/3. The example about using it for password looks realistic to me (though `password` is missing from the example environment and `/controller.inc` is a wrong path).

4. Rspamd will process this file on startup (given the appropriate flag); there is no need for `rspamadm template` unless you want to look at your configs.

5. It's meant to highlight security considerations about who can author this file. Commonly rspamd does _not_ start as root but on some platforms/configurations maybe it does - and if so the `main` process continues to run as root - so it can do on `reload` everything it would do on a normal startup.

6. see 0

7. I suppose once we're already processing config might be too late.

**Reply by TonyGravagno** [reply]
:bulb: A light bulb when on as I was going through this. The text says:
> You can also add values, not merely plain strings but any Lua objects,
> like tables, by specifying additional environment files with the
> --lua-env command line argument.

My thought when I first read that was "what command?" So I looked for "--lua-env" in the doc and the only reference I found was under rspamadm, which has a template command, which has a --lua-env option. Since I'm working on the configuration doc in the Jinja templating section, I thought I found all of the info available, that this was the only mechanism related to Lua setting of environment variables, that --lua-env was limited to the unmentioned rspamadm utility and its template command.

I did not know that the above text was talking about `/usr/bin/rspam`d, **and** `rspamadm`. A [search for "lua-env" in the source](https://github.com/rspamd/rspamd/search?q=lua-env) shows the option is available on both the rspamd executable and the rspamadm executable.

I get this now! When starting the rspamd service we can modify the command-line to include the --lua-env option which will use Lua code to set environment variables. In Debian/Ubuntu we start services with `systemctl start rspamd`. It gets the command-line from /lib/systemd/system/rspamd.service (see `systemctl status rspamd` where it shows "Loaded"). In that file we see 
`ExecStart=/usr/bin/rspamd -c /etc/rspamd/rspamd.conf -f`
This is the command-line that the doc is talking about, where we'd add --lua-env to set environment variables.

I believe that directive is processed before the .conf files are processed, so if we use Lua to set the environment, the conf files will use those variables. In this way, I believe we can override $LOGDIR, $WWWDIR, and other settings - with a high level of caution.

The `rspamadm` utility with its template command also supports --lua-env, so that we can recreate the same environment in a user CLI process that is used for the rspamd.service.

If all of this is correct then I'll transfer this understanding into the doc - and add a new page for /usr/bin/rspamd itself and related options. I will have some follow-ups on this, but for now, am I on the right track?

Thanks!

**Comment by TonyGravagno** [comment]
As a follow-up, now with a better understanding: I see in [the code](https://github.com/rspamd/rspamd/blob/master/cmake/Paths.cmake) where the environment varaiables are set in the build/make process. But I don't see how those values are set for the runtime environment, accessible by $NAME. There is [this code](https://github.com/rspamd/rspamd/blob/master/src/libserver/cfg_utils.c#L1193) in cfg_utils.c, where [ucl_parser_register_variable](https://github.com/rspamd/rspamd/blob/master/contrib/libucl/ucl_parser.c#L2794) is called for each $NAME, but I'm still not connecting how those NAME values are set as $variables.

This is relevant to the docs where it says (no link for now, I was recently looking at it...) that we can only change the default env vars with a rebuild from source. OK, that makes sense, but I'd like to point to the make file as an indication of exactly where to make such a (drastic) change.

Also, to the point and on topic with this thread ... since Lua can set environment variables, can we infer here that we do not need to rebuild to set a varible like $LOGDIR, just invoke --lua-env /some/file.lua from the rspamd (service) command-line?

**Reply by TonyGravagno** [reply]
And to link related material : [This note](https://github.com/rspamd/rspamd/issues/1609#issuecomment-298195493) by @fatalbanana documents yet another way to set the environment.

> You can indeed just run something like
> CONFDIR=/etc/my DBDIR=/var/my rspamd -c /etc/my/rspamd.conf.
> It's not a recommendable way to use rspamd. Settings module should
> obviate need for multiple instances and test instances would be better
> running on dedicated machines/VMs.

With regard to my other notes here, should we attempt to put this in /lib/systemd/system/rspamd.service ? :

`ExecStart=CONFDIR=/etc/my DBDIR=/var/my rspamd -c /etc/my/rspamd.conf`

I don't know how the syntax is processed there, if quotes are required, etc.

Or at this point should we use something like `ExecStart=/local/dir/start_rspamd.sh` ?

**Comment by fatalbanana** [comment]
It's better not to modify systemd units directly & instead create overlay files in `/etc/`.

e.g.
~~~
# Create following file at /etc/systemd/system/rspamd.service.d/lua_env.conf
[Service]
# Clears the existing setting
ExecStart=
# Sets it to something else
ExecStart=/usr/bin/rspamd -c /etc/rspamd/rspamd.conf -f --lua-env /etc/rspamd/lua.env
# Now we can do `systemctl daemon-reload` and `systemctl restart rspamd` to effect the changes
~~~

To set environment variables one wants `EnvironmentFile=` or `Environment=` [directives](https://www.freedesktop.org/software/systemd/man/systemd.exec.html).
